flixel/input/actions/FlxActionSet.hx

```diff
@@ -0,0 +1,322 @@
+		set.steamHandle = FlxSteamController.getActionSetHandle(SteamSet.name);
+		
+		return set;
+	}
+	#end
+	
+	/**
+	 * Create an action set from a parsed JSON object
+	 * 
+	 * @param	Data	A parsed JSON object
+	 * @param	CallbackDigital	A function to call when digital actions fire
+	 * @param	CallbackAnalog	A function to call when analog actions fire
+	 * @return	An action set
+	 */
+	@:access(flixel.input.actions.FlxActionManager)
+	private static function fromJSON(Data:Dynamic, CallbackDigital:FlxActionDigital->Void, CallbackAnalog:FlxActionAnalog->Void):FlxActionSet
+	{
+		var digitalActions:Array<FlxActionDigital> = [];
+		var analogActions:Array<FlxActionAnalog> = [];
+		
+		if (Data == null) return null;
+		
+		if (Reflect.hasField(Data, "digitalActions"))
+		{
+			var arrD:Array<Dynamic> = Reflect.field(Data, "digitalActions");
+			for (d in arrD)
+			{
+				var dName:String = cast d;
+				var action = new FlxActionDigital(dName, CallbackDigital);
+				digitalActions.push(action);
+			}
+		}
+		
+		if (Reflect.hasField(Data, "analogActions"))
+		{
+			var arrA:Array<Dynamic> = Reflect.field(Data, "analogActions");
+			for (a in arrA)
+			{
+				var aName:String = cast a;
+				var action = new FlxActionAnalog(aName, CallbackAnalog);
+				analogActions.push(action);
+			}
+		}
+		
+		if (Reflect.hasField(Data, "name"))
+		{
+			var name:String = cast Reflect.field(Data, "name");
+			var set = new FlxActionSet(name, digitalActions, analogActions);
+			return set;
+		}
+		
+		return null;
+	}
+	
+	public function toJSON():String
+	{
+		var space:String = "\t";
+		return Json.stringify(this, function(key:Dynamic, value:Dynamic):Dynamic{
+			
+			if (Std.is(value, FlxAction))
+			{
+				var fa:FlxAction = cast value;
+				return {
+					"type": fa.type,
+					"name": fa.name,
+					"steamHandle": fa.steamHandle
+				}
+			}
+			
+			return value;
+			
+		}, space);
+	}
+	
+	public function new(Name:String, DigitalActions:Array<FlxActionDigital>, AnalogActions:Array<FlxActionAnalog>)
+	{
+		name = Name;
+		digitalActions = DigitalActions;
+		analogActions = AnalogActions;
+	}
+	
+	/**
+	 * Automatically adds or removes inputs for a steam controller
+	 * to any steam-affiliated actions
+	 * @param	Handle	steam controller handle from FlxSteam.getConnectedControllers(), or FlxInputDeviceID.FIRST_ACTIVE / ALL
+	 * @param	Attach	true: adds inputs, false: removes inputs
+	 */
+	public function attachSteamController(Handle:Int, Attach:Bool = true):Void
+	{
+		attachSteamControllerSub(Handle, Attach, FlxInputType.DIGITAL, digitalActions, null);
+		attachSteamControllerSub(Handle, Attach, FlxInputType.ANALOG, null, analogActions);
+	}
+	
+	/**
+	 * Add a digital action to this set if it doesn't already exist
+	 * @param	Action a FlxActionDigital
+	 * @return	whether it was added
+	 */
+	public function addDigital(Action:FlxActionDigital):Bool
+	{
+		if (digitalActions.indexOf(Action) != -1) return false;
```

Comment:
(Via static extension though)
Am 14.05.2016 13:36 schrieb "Zaphod" <notifications@github.com>:

> In flixel/input/actions/FlxActionSet.hx
> <https://github.com/HaxeFlixel/flixel/pull/1805#discussion_r63274929>:
>
> > +	 * @param	Attach	true: adds inputs, false: removes inputs
> > +	 */
> > +	public function attachSteamController(Handle:Int, Attach:Bool = true):Void
> > +	{
> > +		attachSteamControllerSub(Handle, Attach, FlxInputType.DIGITAL, digitalActions, null);
> > +		attachSteamControllerSub(Handle, Attach, FlxInputType.ANALOG, null, analogActions);
> > +	}
> > +	
> > +	/**
> > +	 * Add a digital action to this set if it doesn't already exist
> > +	 * @param	Action a FlxActionDigital
> > +	 * @return	whether it was added
> > +	 */
> > +	public function addDigital(Action:FlxActionDigital):Bool
> > +	{
> > +		if (digitalActions.indexOf(Action) != -1) return false;
>
> as @Gama11 <https://github.com/Gama11> mentioned you can change this line
> to:
>
> if (FlxArrayUtil.contains(digitalActions, Action)) return false;
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFlixel/flixel/pull/1805/files/d62bf2c0c11d1dec8da3c05d5c29fbac247ef1ad#r63274929>
>